### PR TITLE
Split API servers from integration tests

### PR DESCRIPTION
**Pull request by soltysh**
This replaces the old setup mechanism with the new API startup for the integration tests.

Current list of failing integration tests:
- [ ] ~~TestAuthProxyOnAuthorize~~ disabled and tracked in: https://bugzilla.redhat.com/show_bug.cgi?id=1640351 
- [x] TestBasicGroupManipulation
- [ ] ~~TestBuildRunningPodDeleteController~~ disabled and tracked in https://bugzilla.redhat.com/show_bug.cgi?id=1641186
- [x] TestClusterQuota
- [ ] ~~TestDNS~~ removed and tracked in: https://bugzilla.redhat.com/show_bug.cgi?id=1639593
- [x] TestEtcd3StoragePath - updated and migration tracked in https://bugzilla.redhat.com/show_bug.cgi?id=1640292
- [ ] ~~TestFrontProxy~~ disabled and tracked in: https://bugzilla.redhat.com/show_bug.cgi?id=1640351
- [ ] ~~TestKubernetesAdmissionPluginConfigFile~~ (removed)
- [ ] ~~TestKubernetesAdmissionPluginEmbeddedConfig~~ (removed)
- [ ] ~~TestKubernetesAdmissionPluginOrderOverride~~ (removed)
- [ ] ~~TestOadmPodNetwork~~ removed and tracked in https://bugzilla.redhat.com/show_bug.cgi?id=1641637
- [x] TestOAuthBasicAuthPassword
- [ ] ~~TestOAuthOIDC~~ disabled and tracked in: https://bugzilla.redhat.com/show_bug.cgi?id=1640351
- [ ] ~~TestOAuthRequestHeader~~ disabled and tracked in: https://bugzilla.redhat.com/show_bug.cgi?id=1640351
- [ ] ~~TestOpenshiftAdmissionPluginConfigFile~~ (removed)
- [ ] ~~TestOpenshiftAdmissionPluginEmbeddedConfig~~ (removed)
- [ ] ~~TestOpenshiftAdmissionPluginOrderOverride~~ (removed)
- [x] TestRestrictUsers
- [x] TestRootRedirect
- [x] TestSNI
- [x] TestStorageVersions
- [ ] ~~TestWebhookTokenAuthn~~ disabled and tracked in: https://bugzilla.redhat.com/show_bug.cgi?id=1640351

### CODE REVIEW

**Comment by soltysh** [comment]
Rebased and built on top of https://github.com/openshift/origin/pull/21175

**Comment by soltysh** [comment]
Weird, that integration failure works fine for me locally. I've rebased and re-pushed it again. Let's see what happens.

**Comment by deads2k** [comment]
wow, just one test `TestFrontProxy`?

If that is it, skip the test and make a bugzilla for @openshift/sig-security

File: test/integration/api_paths_test.go [hunk 1]
```diff
@@ -48,25 +44,26 @@ func TestRootAPIPaths(t *testing.T) {
...
 	// any not specified are not
-	expectedOpenShiftAPILevels := sets.NewString(masterConfig.APILevels...)
-	expectedKubeAPILevels := sets.NewString(configapi.GetEnabledAPIVersionsForGroup(masterConfig.KubernetesMasterConfig, kapi.GroupName)...)
-	actualOpenShiftAPILevels := sets.String{}
-	actualKubeAPILevels := sets.String{}
-	for _, route := range broadcastRootPaths.Paths {
-		if strings.HasPrefix(route, "/oapi/") {
-			actualOpenShiftAPILevels.Insert(route[6:])
-		}
+	// TODO: the below test doesn't make sense anymore
```
Comment (deads2k) [inline test/integration/api_paths_test.go]: Agree.  Delete the entire test

**Comment by soltysh** [comment]
/hold
until I debug

File: pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/flags.go [hunk 2]
```diff
@@ -251,6 +250,7 @@ func admissionFlags(admissionPluginConfig map[string]configv1.AdmissionPluginCon
 
 		} else {
 			forceOn = append(forceOn, pluginName)
+			pluginConfig[pluginName] = config
```
Comment (soltysh) [inline pkg/cmd/openshift-kube-apiserver/openshiftkubeapiserver/flags.go]: @deads2k does this make sense?

**Comment by deads2k** [comment]
The admission configuration paths don't exist anymore.  Remove those integration tests.

**Comment by deads2k** [comment]
kill TestDNS, write a bugzilla and assign to networking.

**Comment by deads2k** [comment]
Talk to @enj and @ericavonb  about disabling the oauth config tests. My memory says they work by starting "fake" apiservers, so they need to final alternatives.

I would expect `TestRestrictUsers` to work.  `TestRootRedirect` can be removed.

File: test/integration/dns_test.go [hunk 3]
```diff
@@ -1,331 +1,331 @@
 package integration
 
-import (
-	"fmt"
-	"hash/fnv"
-	"net"
-	"strconv"
-	"testing"
-	"time"
+// import (
```
Comment (deads2k) [inline test/integration/dns_test.go]: delete this entire file in one commit.  Write a high severity bugzilla referencing it for @ironcladlou .

The apiservers no longer include dns.

File: test/integration/etcd_storage_path_test.go [hunk 4]
```diff
@@ -497,6 +497,7 @@ var etcdStorageData = map[schema.GroupVersionResource]struct {
 	gvr("apiregistration.k8s.io", "v1beta1", "apiservices"): {
 		stub:             `{"metadata": {"name": "as1.foo.com"}, "spec": {"group": "foo.com", "version": "as1", "groupPriorityMinimum":100, "versionPriority":10}}`,
 		expectedEtcdPath: "kubernetes.io/apiservices/as1.foo.com",
+		expectedGVK:      gvkP("apiregistration.k8s.io", "v1", "APIService"),
```
Comment (deads2k) [inline test/integration/etcd_storage_path_test.go]: write a high severity bugzilla for the master team to fix apiservice and SCC in 4.0 for migration.

Comment (deads2k) [reply test/integration/etcd_storage_path_test.go]: > write a high severity bugzilla for the master team to fix apiservice and SCC in 4.0 for migration.

for the path I mean.  This triggered my memory.

File: test/util/server/server.go [hunk 5]
```diff
@@ -57,6 +67,21 @@ var (
 	startedMaster bool
 	// startedNode is true if the node has already been started in process
 	startedNode bool
+
+	openshiftGVs = []schema.GroupVersion{
+		{Group: "apps.openshift.io", Version: "v1"},
+		{Group: "authorization.openshift.io", Version: "v1"},
+		{Group: "build.openshift.io", Version: "v1"},
+		{Group: "image.openshift.io", Version: "v1"},
+		{Group: "network.openshift.io", Version: "v1"},
```
Comment (deads2k) [inline test/util/server/server.go]: remove this.  We no longer support it.

**Comment by deads2k** [comment]
I'd like to know if `TestClusterQuota` is still outstanding. I think separate commits split on teams for various skips so they can easily revert paired with high severity bugzillas and links here is suitable to finish the split.

File: test/integration/etcd_storage_path_test.go (hunk 4 above)
Comment (soltysh) [reply test/integration/etcd_storage_path_test.go]: Created https://bugzilla.redhat.com/show_bug.cgi?id=1640292.

File: test/util/server/server.go (hunk 5 above)
Comment (soltysh) [reply test/util/server/server.go]: Done.

**Comment by soltysh** [comment]
/hold cancel
All the bits are in place, the disabled tests are tracked in linked BZ. Let's now try to figure out how our CI will react to this.

File: test/integration/auth_proxy_test.go [hunk 6]
```diff
@@ -1,121 +1,121 @@
...
-		t.Logf("Going to %v\n", req.URL)
-		*redirectRecord = append(*redirectRecord, *req.URL)
-
-		if len(via) >= 10 {
-			return errors.New("stopped after 10 redirects")
-		}
-		return nil
-	}
-}
+// import (
```
Comment (enj) [inline test/integration/auth_proxy_test.go]: Please use `t.Skip()` to disable these instead of commenting them out.

**Comment by soltysh** [comment]
Rebased, skipped tests as @enj wanted let's try getting that integration green.

**Comment by soltysh** [comment]
Got blessing from David, tagging.

**Pull request merged**